### PR TITLE
feat(shoplist): add aggregator for base-unit ingredient summation

### PR DESCRIPTION
**Pull request by Katsiarynakavaleuskaya**
Adds offline aggregator that groups IngredientSpec by food_id and sums quantities. Works only with base units (G, ML, PCS); raises ValueError for non-base units or unit mismatches.

- aggregate_specs: groups by food_id, sums quantities
- Validates base units only (G, ML, PCS)
- Detects unit mismatches for same food_id
- Deterministic output (sorted by food_id)
- Preserves first FoodRef instance per food_id
- Zero quantities allowed and preserved

All functions are pure (no I/O, env, time, random dependencies). Anchor tests ensure invariants are preserved.

# Pull Request

## Summary

(Free-form description of changes, e.g., "Fix crash in BMI calculation when height is zero")

- [ ] Select one change type:
  - [ ] Bug fix
  - [ ] Feature
  - [ ] Refactor
  - [ ] Docs
- [ ] Linked issues/PRs: #

## Risk & Impact

(User-facing change? Data model/migration? Security- or Performance-sensitive?)

- [ ] User-facing change
- [ ] Data model/migration
- [ ] Security-sensitive
- [ ] Performance-sensitive

## Test Plan

(Unit: small isolated functions; Integration: endpoints/DB; Manual: steps to verify)

- [ ] Unit tests updated/added
- [ ] Integration/slow tests (if applicable)
- [ ] Manual verification steps

## CI Gates

(PR tests must pass; Diff coverage means tests cover changed lines ≥ threshold)

- [ ] PR tests green (lint, type, unit)
- [ ] Diff coverage ≥ 97% on changed lines

### For Simple Changes

Use this checklist to confirm the PR is truly simple:

- [ ] No database/schema/migration changes
- [ ] No public API contract changes (endpoints, request/response, events)
- [ ] Covered by existing tests (or adds ≤ 1-2 focused unit tests)
- [ ] < 50 LOC changed (excluding tests/docs)
- [ ] No performance or security impact

Example: Copy change in docs, minor log level tweak, small refactor of a pure function.

Rollback / Feature flag (brief):

- How to revert: describe the commit to revert or config to change
- Feature flag/toggle (if any): name and how to disable
- Owner for emergency contact: @username

### For Complex/High-Risk Changes

Please fill out the following sections if applicable:

#### Deployment Strategy

- [ ] Deployment order for multi-service changes (if services depend on each other)
- [ ] Feature flag configuration (enable/disable without redeploy)
- [ ] Blue-green / canary deployment plan (if applicable)

#### Database & Data Changes

- [ ] Database migrations required (Alembic version, backwards compatibility)
- [ ] Data migration/backfill steps (scripts, rollback procedures)
- [ ] Data cleanup steps (if removing deprecated data)
- [ ] Backwards compatibility guarantees (old clients still work)

#### Monitoring & Observability

- [ ] New monitoring/alerting rules to add (metrics, thresholds, SLOs)
- [ ] Dashboards to create/update (Grafana, DataDog, etc.)
- [ ] Logging changes (new log levels, structured logs, correlation IDs)
- [ ] Health check endpoints affected

#### Post-Deploy Verification

- [ ] Manual verification checklist (specific endpoints, user flows)
- [ ] Smoke tests to run (automated or manual)
- [ ] Performance benchmarks to verify (latency, throughput)
- [ ] Rollback triggers (what conditions require immediate rollback)

#### Additional Context

- [ ] Breaking changes (API contracts, response formats)
- [ ] Dependencies updated (requirements.txt, package versions)
- [ ] Configuration changes (env vars, secrets, feature toggles)
- [ ] Documentation updates needed (README, API docs, runbooks)

## Summary by Sourcery

Добавлен чистый агрегатор для объединения нормализованных спецификаций ингредиентов в детерминированные строки списка покупок, сгруппированные по `food_id` и суммированные в базовых единицах.

New Features:
- Предоставлен вспомогательный метод `aggregate_specs`, который группирует элементы `IngredientSpec` по `food_id` и формирует итоговые значения `ShoplistLine`, используя только базовые единицы.

Tests:
- Добавлены комплексные опорные тесты, проверяющие поведение агрегации, валидацию единиц измерения, детерминизм и крайние случаи, такие как нулевые и дробные количества.

<details>
<summary>Original summary in English</summary>

## Summary by Sourcery

Add a pure aggregator to combine normalized ingredient specifications into deterministic shopping list lines grouped by food_id and summed in base units.

New Features:
- Expose an aggregate_specs helper that groups IngredientSpec items by food_id and produces ShoplistLine totals using base units only.

Tests:
- Add comprehensive anchor tests covering aggregation behavior, unit validation, determinism, and edge cases such as zero and fractional quantities.

</details>

## Summary by CodeRabbit

* **New Features**
  * Added an ingredient aggregation capability that groups items by food and combines quantities into single lines.
  * Enforces use of standard base units (g, ml, pcs) and rejects unsupported or mixed units.
  * Returns deterministic, consistently sorted results.

* **Tests**
  * Added comprehensive tests covering aggregation behavior, unit validation, ordering, and edge cases.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: tests/core/test_shoplist_aggregator.py [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+        ]
+        lines = aggregate_specs(specs)
+        assert len(lines) == 3
+        # Should be sorted alphabetically
+        assert lines[0].food.food_id == "apple"
+        assert lines[1].food.food_id == "banana"
+        assert lines[2].food.food_id == "zebra"
+
+
+class TestAggregateSpecsEdgeCases:
```
[the remark this answers is not in the document]
Comment (Katsiarynakavaleuskaya) [reply tests/core/test_shoplist_aggregator.py:247]: Thanks for the suggestions!

**Negative quantities**
Negative values are intentionally disallowed at the domain level.
`Quantity` enforces `value >= 0` during construction, so `aggregate_specs`
never receives negative quantities by design. The aggregator operates only
on valid domain objects and is not a boundary for raw/unclean input, so we
intentionally do not add duck-typed negative tests here.

**Immutability**
`IngredientSpec`, `Quantity`, and `FoodRef` are frozen dataclasses.
Immutability is guaranteed by the type system itself, not by runtime behavior
of the aggregator. Since mutation is structurally impossible without bypassing
the domain model, we intentionally do not duplicate this guarantee with a
snapshot-based test.

Both behaviors are considered fixed domain invariants.
